File: src/optical_element/MATLAB/Furion_slits.m
classdef Furion_slits<handle
    %¾ØÐÎslits
    
    properties
        beam_out;%slitsÊä³ö¹âÊø
    end
    
    methods 
        function obj = Furion_slits(beam_in,center_x,center_y,width,height) 
            xx = beam_in.X(:);
            yy = beam_in.Y(:);
            vv = beam_in.field(:);
            n = length(xx);
            N = beam_in.N;
       % coord =zeros(1,n);
        for i = 1:n
            if ~(abs(xx(i)-center_x)<width/2 && abs(yy(i)-center_y)<height/2)
                vv(i) =0;
                i
            end
        end
       X=reshape(xx,[N,N]);
       Y=reshape(yy,[N,N]);
       V =reshape(vv,[N,N]);
        %k=find(coord);
        obj.beam_out = beam(X,Y,V,beam_in.wavelength);
        end
    end
    
end

File: src/optical_element/Furion_slits.cpp
#include "Furion_slits.h"

using namespace Furion_NS;

Furion_Slits::Furion_Slits()
{
	cout << "Furion_Slits ³õÊ¼»¯" << endl;
}

Furion_Slits::~Furion_Slits()
{

}

void Furion_Slits::Furion_slits(Beam* beam_in, double center_x, double center_y, double width, double height, int N)
{
    std::vector<std::vector<std::complex<double>> > V;
	for (int i = 0; i < N; i++)
	{
		for (int j = 0; j < N; j++)
		{
			V[i][j] = std::complex<double>(!(std::abs(beam_in->X[i][j] - center_x) < width / 2 && std::abs(beam_in->Y[i][j] - center_y) < height / 2), 0);
		}
	}

	beam_out = new Beam(beam_in->X, beam_in->Y, V, beam_in->wavelength, beam_in->N);
}
